### PR TITLE
JCLOUDS-651: Add copyBlob to portable abstraction and add provider optimized variants

### PR DESCRIPTION
**Pull request by andrewgaul**

### CODE REVIEW

**Comment by everett-toews** [comment]
Thanks for the pull request but it's release week in jclouds and that means it's time to clean up the PR queue. This PR will be over 6 months old as of April 1. If you intend to continue work on it, please make a comment by April 2. Otherwise it will be closed on April 3.

File: blobstore/src/main/java/org/jclouds/blobstore/BlobStore.java [hunk 1]
```diff
@@ -224,6 +227,16 @@
    String putBlob(String container, Blob blob, PutOptions options);
 
    /**
+    * Copy blob from one container to another.  Some providers implement this
+    * more efficiently than corresponding getBlob and putBlob operations.
```
Comment (andrewgaul) [inline blobstore/src/main/java/org/jclouds/blobstore/BlobStore.java]: Done.

File: blobstore/src/main/java/org/jclouds/blobstore/AsyncBlobStore.java [hunk 2]
```diff
@@ -138,6 +140,12 @@
    ListenableFuture<String> putBlob(String container, Blob blob, PutOptions options);
 
    /**
+    * @see BlobStore#copyBlob(String,String,String,String)
+    */
+   ListenableFuture<String> copyBlob(String fromContainer, String fromName, String toContainer, String toName,
+         Optional<CopyOptions> options);
```
Comment (andrewgaul) [inline blobstore/src/main/java/org/jclouds/blobstore/AsyncBlobStore.java]: Done.

**Comment by andrewgaul** [comment]
I added partial support for openstack-swift and reported [JCLOUDS-872](https://issues.apache.org/jira/browse/JCLOUDS-872) for the remaining features.  Unfortunately Swift itself does not have the conditional copy options that other providers offer.  We may want to implement this with two Swift RPCs, e.g., `blobMetadata`, check lastModified/ETag in jclouds, then call `copyBlob`, to gain cross-provider support.

**Comment by andrewgaul** [comment]
The current implementation only allows overriding user metadata, not system metadata like Content-Type.  Debating whether to push as-is or wait for some other provider support like [JCLOUDS-872](https://issues.apache.org/jira/browse/JCLOUDS-872) and [JCLOUDS-875](https://issues.apache.org/jira/browse/JCLOUDS-875) to land first.

**Comment by andrewgaul** [comment]
I will address system metadata in subsequent commits.

**Pull request merged**